cmd/replay/main.go

```diff
@@ -283,6 +285,62 @@ func main() {
 		batchFetcher := func(batchNum uint64) ([]byte, error) {
 			return wavmio.ReadInboxMessage(batchNum), nil
 		}
+
+		// Handle the various pre-conditions and panics that can happen before we should enter the espresso logic in the validators STF
+		validatingAgainstEspresso, panicHandler := handleEspressoPreConditions(message, chainConfig.ArbitrumChainParams.EnableEspresso)
+		if validatingAgainstEspresso {
+			txs, jst, err := arbos.ParseEspressoMsg(message.Message)
+			if err != nil {
+				panic(err)
+			}
+			if jst == nil {
+				panic("batch missing espresso justification")
+			}
+
+			hotshotHeader := jst.Header
+			height := hotshotHeader.Height
+
+			// Check the continuity of the hotshot block if we are not running the sovereign sequencer.
+			if !arbos.IsEspressoSovereignMsg(message.Message) {
+				validatedHeight := wavmio.GetEspressoHeight()
+				if validatedHeight == 0 {
+					// Validators can choose their own trusted starting point to start their validation.
+					// TODO: Check the starting point is greater than the first valid hotshot block number.
+					wavmio.SetEspressoHeight(height)
+				} else if validatedHeight+1 == height {
+					wavmio.SetEspressoHeight(height)
+				} else {
+					panic(fmt.Sprintf("invalid hotshot block height: %v, got: %v", height, validatedHeight+1))
+				}
+			}
+			if jst.BlockMerkleJustification == nil {
+				panic("block merkle justification missing")
+			}
+
+			if jst.Proof == nil && len(txs) != 0 {
+				panic("namespace proof missing")
+			}
+
+			commitment := espressoTypes.Commitment(wavmio.ReadHotShotCommitment(height))
+			jsonHeader, err := json.Marshal(hotshotHeader)
+			if err != nil {
+				panic("unable to serialize header")
+			}
+			espressocrypto.VerifyMerkleProof(
+				jst.BlockMerkleJustification.BlockMerkleProof.Proof,
+				jsonHeader,
+				*jst.BlockMerkleJustification.BlockMerkleComm,
+				commitment,
+			)
+			if jst.Proof != nil {
+				espressocrypto.VerifyNamespace(chainConfig.ChainID.Uint64(), *jst.Proof, *jst.Header.PayloadCommitment, *jst.Header.NsTable, txs, *jst.VidCommon)
```

Comment:
This call dispatches to the espresso crypto helper to verify the integrity of the Espresso namespace. Currently this checks that the L2 transactions are exactly the ones that were finalized by Espresso consensus.